src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/CanonicalMiscFilesProjectLoader.cs

```diff
@@ -0,0 +1,317 @@
+        }, cancellationToken);
+
+        Contract.ThrowIfNull(addedDocument);
+        return addedDocument;
+    }
+
+    private async ValueTask<TextDocument> AddDocumentToPrimordialProjectAsync(string documentPath, SourceText documentText, ProjectId primordialProjectId, CancellationToken cancellationToken)
+    {
+        var canonicalDocumentPath = _canonicalDocumentPath.Value;
+        var miscWorkspace = _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.Workspace;
+
+        var documentInfo = DocumentInfo.Create(
+            DocumentId.CreateNewId(primordialProjectId),
+            name: Path.GetFileName(documentPath),
+            loader: TextLoader.From(TextAndVersion.Create(documentText, VersionStamp.Create())),
+            filePath: documentPath);
+
+        await _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.ApplyChangeToWorkspaceAsync(workspace =>
+        {
+            workspace.OnDocumentAdded(documentInfo);
+        }, cancellationToken);
+
+        var addedDocument = miscWorkspace.CurrentSolution.GetRequiredDocument(documentInfo.Id);
+        return addedDocument;
+    }
+
+    private TextDocument CreatePrimordialProjectAndAddDocument(string documentPath, SourceText documentText)
+    {
+        var canonicalDocumentPath = _canonicalDocumentPath.Value;
+
+        // Create primordial project with the canonical document
+        var canonicalText = SourceText.From(string.Empty);
+        var canonicalLoader = new SourceTextLoader(canonicalText, canonicalDocumentPath);
+
+        var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
+            _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.Workspace,
+            canonicalDocumentPath,
+            canonicalLoader,
+            new LanguageInformation(LanguageNames.CSharp, scriptExtension: null),
+            canonicalText.ChecksumAlgorithm,
+            _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.Workspace.Services.SolutionServices,
+            metadataReferences: []);
+
+        // Add the project first, then add the requested document
+        _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.ApplyChangeToWorkspace(workspace =>
+        {
+            workspace.OnProjectAdded(projectInfo);
+        });
+
+        // Now add the requested document
+        var documentInfo = DocumentInfo.Create(
+            DocumentId.CreateNewId(projectInfo.Id),
+            name: Path.GetFileName(documentPath),
+            loader: new SourceTextLoader(documentText, documentPath),
+            filePath: documentPath);
+
+        _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.ApplyChangeToWorkspace(workspace =>
+        {
+            workspace.OnDocumentAdded(documentInfo);
+        });
+
+        // Begin loading the canonical project with a design-time build
+        BeginLoadingProjectWithPrimordial_NoLock(
```

Comment:
It doesn't look like a lock is held while we are making this call. If it is expected that the caller of `CreatePrimordialProjectAndAddDocument` holds a lock, then perhaps add the `_NoLock` suffix to that.